tools/importer-msgraph-metadata/pipeline/task_parse_resources.go

```diff
@@ -0,0 +1,294 @@
+package pipeline
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+func (p pipelineTask) parseResourcesForService(resourceIds ResourceIds, models Models) (resources Resources, err error) {
+	resources = make(Resources)
+	for pathKey, pathItem := range p.spec.Paths {
+		path := strings.Clone(pathKey)
+		operations := pathItem.Operations()
+		operationTags := make([]string, 0)
+
+		// Check tags and skip
+		skip := true
+		for _, operation := range operations {
+			if tagMatches(p.service, operation.Tags) {
+				operationTags = append(operationTags, operation.Tags...)
+				skip = false
+				break
+			}
+		}
+		if skip {
+			continue
+		}
+
+		parsedPath := NewResourceId(path, operationTags)
+		lastSegment := parsedPath.Segments[len(parsedPath.Segments)-1]
+		if lastSegment.Type == SegmentCast || lastSegment.Type == SegmentFunction {
+			p.logger.Info(fmt.Sprintf("skipping path containing cast or function for %q: %v", p.service, path))
+			continue
+		}
+
+		resourceName := ""
+		if r, ok := parsedPath.FindResourceName(); ok {
+			resourceName = *r
+		}
+		if resourceName == "" {
+			p.logger.Warn(fmt.Sprintf("path with unknown name was encountered for %q: %v", p.service, path))
+			continue
+		}
+
+		fullyQualifiedResourceName := ""
+		if fqrn, ok := parsedPath.FullyQualifiedResourceName(); ok {
+			fullyQualifiedResourceName = *fqrn
+		}
+		if fullyQualifiedResourceName == "" {
+			p.logger.Warn(fmt.Sprintf("path with unknown fully-qualified name was encountered for %q: %v", p.service, path))
+			continue
+		}
+
+		// Resources by default go into their own category when their final URI segment is a label
+		resourceCategory := ""
+		if lastSegment.Type == SegmentLabel || lastSegment.Type == SegmentUserValue {
+			resourceCategory = fullyQualifiedResourceName
+		}
+
+		if _, ok := resources[fullyQualifiedResourceName]; !ok {
+			// Create a new resource if not already encountered
+			p.logger.Info(fmt.Sprintf("found new resource %q (category %q, service %q, version %q)", resourceName, resourceCategory, p.service, p.apiVersion))
+
+			resources[fullyQualifiedResourceName] = &Resource{
+				Name:       resourceName,
+				Category:   resourceCategory,
+				Version:    p.apiVersion,
+				Service:    cleanName(p.service),
+				Paths:      []ResourceId{parsedPath},
+				Operations: make([]Operation, 0, len(operations)),
```

Comment:
given we're skipping operations (line 78 below) - this'd lead to potential `nil` items at the end, fwiw?

```suggestion
				Operations: make([]Operation, 0),
```